src/oaklib/implementations/ontoportal/ontoportal_implementation_base.py

```diff
@@ -0,0 +1,271 @@
+import logging
+from abc import ABC
+from dataclasses import dataclass, field
+from typing import Any, ClassVar, Dict, Iterable, Iterator, List, Tuple, Union
+from urllib.parse import quote
+
+import requests
+from ontoportal_client.api import PreconfiguredOntoPortalClient
+from sssom_schema import Mapping
+
+from oaklib.datamodels.search import SearchConfiguration
+from oaklib.datamodels.text_annotator import TextAnnotation, TextAnnotationConfiguration
+from oaklib.datamodels.vocabulary import SEMAPV
+from oaklib.interfaces import (
+    MappingProviderInterface,
+    SearchInterface,
+    TextAnnotatorInterface,
+)
+from oaklib.interfaces.basic_ontology_interface import METADATA_MAP, PREFIX_MAP
+from oaklib.types import CURIE, URI
+from oaklib.utilities.apikey_manager import get_apikey_value
+from oaklib.utilities.rate_limiter import check_limit
+
+SEARCH_CONFIG = SearchConfiguration()
+
+# See:
+#   https://www.bioontology.org/wiki/BioPortal_Mappings
+#   https://github.com/agroportal/project-management/wiki/Mappings
+SOURCE_TO_PREDICATE = {
+    "CUI": "skos:closeMatch",
+    "LOOM": "skos:closeMatch",
+    "REST": "skos:relatedMatch",  # maybe??
+    "SAME_URI": "skos:exactMatch",
+}
+
+
+@dataclass
+class OntoPortalImplementationBase(
+    TextAnnotatorInterface, SearchInterface, MappingProviderInterface, ABC
+):
+
+    OntoPortalClientClass: ClassVar[type[PreconfiguredOntoPortalClient]] = None
+    api_key_name: ClassVar[str] = None
+
+    label_cache: Dict[CURIE, str] = field(default_factory=lambda: {})
+    ontology_cache: Dict[URI, str] = field(default_factory=lambda: {})
+    focus_ontology: str = None
+
+    def __post_init__(self):
+        if self.focus_ontology is None:
+            if self.resource:
+                self.focus_ontology = self.resource.slug
+        if not self.OntoPortalClientClass:
+            raise NotImplementedError("OntoPortalClientClass not specified")
+        api_key = get_apikey_value(self.api_key_name)
+        self.client = self.OntoPortalClientClass(api_key=api_key)
+
+    def prefix_map(self) -> PREFIX_MAP:
+        # TODO
+        return {}
+
+    def _get_response(self, path: str, *args, raise_for_status=True, **kwargs):
+        check_limit()
+        path = path.removeprefix(self.client.base_url)
+        response = self.client.get_response(path, *args, **kwargs)
+        if raise_for_status:
+            response.raise_for_status()
+        return response
+
+    def _get_json(self, *args, **kwargs):
+        response = self._get_response(*args, **kwargs)
+        return response.json()
+
+    def ontologies(self) -> Iterable[CURIE]:
```

Comment:
the type annotation here for the result is misleading. the acronyms are prefixes, not CURIEs

Reply:
good point, I think this is something that may need a deeper fix as part of a separate PR through

Reply:
I think I'm with Chris on this one. It's not good, but it's also not an isolated thing and would probably be handled better in a separate PR.